### ISSUE TITLE
Make all build dependencies optional

### ISSUE THREAD

**Issue by carterbox** [tomopy/tomopy#428]
This feature is related to the following recent issues with other institutions trying to build TomoPy with the new build system and dependencies:

- https://github.com/NSLS-II/docs/pull/112
- https://github.com/conda-forge/tomopy-feedstock/pull/21
- I also had a request for a no-MKL build (for licensing reasons), but I can't find the Issue.

It seems that people are both not expecting the hard OpenCV and MKL build dependencies, but might not need or want them.

**Describe the solution you would like.**
We should make all of the build dependencies, MKL, OpenCV, and CUDA optional, and return a helpful error if users try to use the algorithms that depend on these libraries which explains that they need to have these libraries installed at build-time.

**Describe alternatives you've considered.**
The alternative is to do nothing.

**Additional context.**
CUDA is already optional at build time, but as described in #422, the CUDA methods will fall back instead of returning an error if CUDA fails.

**Comment by newville** [comment]
@carterbox +1 from me: reducing the dependencies and installing with `pip` would be awesome.

For MKL,  I believe that older versions of gridrec used neither MKL nor FFTW.  It may be possible to have a version that could use fftpack (not the fastest, but included with numpy and scipy as shared libs) if MKL was unavailable.  That would probably be some real work.

**Comment by carterbox** [comment]
I've met a few more people at ICTMS this week that are interested in TomoPy for its preprocessing features, but not reconstruction (because they use fan or come beam). Thus they don't necessarily need the features that MKL and OpenCV provide.

@jrmadsen Do you have time to implement this (Making dependencies optional not replacing MKL FFT)?

**Comment by jrmadsen** [comment]
@carterbox I have made significant progress on a semi-redesign that more-or-less implements a backend scheme for the CUDA/OpenCV implementations. Unfortunately, my GPU clusters are undergoing to maintenance so I won't be able to do full testing until later next week.

Could you handle making MKL optional in the source (and sort out any potential issues)? You can commit directly to the branch on my fork if you want. It should be relatively easy to disable, essentially all you need to do is surround the entire source with the preprocessor:

```cpp
#if defined(TOMOPY_USE_MKL)

// original source...

#else
void gridrec(...)
{
    printf(stderr, "Error! TomoPy did not find MKL during configure time and was thus not compiled "
                  "with MKL support. Without MKL, the gridrec reconstruction algorithm is not "
                  "available. If you would like to be able to utilize gridrec without a dependence "
                  "on MKL, TomoPy encourages contributions from users. "
                  "In order to enable gridrec with MKL, please see the documentation at <link>");
    exit(EXIT_FAILURE);
}
#endif
```

Add this to `cmake/Modules/Options.cmake`: 

```cmake
add_option(TOMOPY_USE_MKL "Enable MKL support" ON)
````

And change this in `cmake/Modules/Packages.cmake`:

```cmake
find_package(MKL REQUIRED)

if(MKL_FOUND)
    list(APPEND EXTERNAL_INCLUDE_DIRS ${MKL_INCLUDE_DIRS})
    list(APPEND EXTERNAL_LIBRARIES ${MKL_LIBRARIES})
endif()
```

to non-required and (for simplicity) create a CMake INTERFACE library that provides the MKL libs/include/compile-defs:

```cmake
find_package(MKL)

if(MKL_FOUND)
    add_library(tomopy-mkl INTERFACE)
    set_target_properties(tomopy-mkl PROPERTIES
        INTERFACE_COMPILE_DEFINITIONS    TOMOPY_USE_MKL
        INTERFACE_INCLUDE_DIRECTORIES    ${MKL_INCLUDE_DIRS}
        INTERFACE_LINK_LIBRARIES         ${MKL_LIBRARIES})
    list(APPEND EXTERNAL_LIBRARIES tomopy-mkl)
endif()
```

@newville There is nothing really stopping TomoPy from being installed via pip, even now, if it were uploaded to PyPi. You just have to install the packages and (contrary to popular belief), conda is not a _Python_ package manager. It can actually be used as a _general_ package manager like the Linux package managers (`apt`, `yum`, etc.). So you can use conda to just install MKL and OpenCV and not install Python. 

For simplicity, I am just going to assume conda installed at `C:\devel\conda\bin\conda.exe` and it is in the PATH and the `cmake` is installed on the system somewhere also in your PATH:

```shell
# use conda to get MKL and OpenCV libraries, no need to install python
conda create -n libraries -c conda-forge mkl mkl-devel libopencv
# append the base directory of these library installations to CMAKE_PREFIX_PATH
$env:CMAKE_PREFIX_PATH = "C:\devel\conda\envs\libraries;$env:CMAKE_PREFIX_PATH"
# probably necessary
pip install scikit-build
# assuming we add to PyPi
pip install -vvv tomopy
```

Also, for anybody else reading this, on Linux, you can skip using conda and follow the instructions [here](https://software.intel.com/en-us/articles/installing-intel-free-libs-and-python-apt-repo) to get MKL via apt (or find similar page for yum) and just do something like: 

```shell
apt-get install libopencv-dev intel-mkl-64bit-2019.4-070 
source /opt/intel/mkl/bin/mklvars.sh intel64 lp64
pip install scikit-build
# assuming we add to PyPi
pip install -vvv tomopy
```

**Comment by newville** [comment]
@jrmadsen  @carterbox 

> @newville There is nothing really stopping TomoPy from being installed via pip, even now, if it were uploaded to PyPi. 

OK.  I think having tomopy on PyPI and readily installable with pip would be helpful to many people.  

> You just have to install the packages and (contrary to popular belief), conda is not a Python package manager. It can actually be used as a general package manager like the Linux package managers (apt, yum, etc.). So you can use conda to just install MKL and OpenCV and not install Python.

Yep. I'm not sure anyone here thinks of `conda` as Python-only.  As an aside: if `scikit-build` is a requirement for installing `tomopy`, you could (probably "should") just list that in the `install_requires` or `setup_requires` argument to `setup()`.


Re: reducing dependencies and MKL for gridrec:  

As you all might know, Numpy 1.7 is now using `pocketfft` which (if I understand the benchmarks) is approximately as fast and accurate as pyfftw (I'm not sure about comparisons to MKL).  In the spirit of reducing dependencies, it might be a good time to explore the possibility of using numpy's C interface to pocketfft for gridrec.  Since numpy is already a hard dependency, using its FFT interface might simplify packaging, testing, and maintenance.  
To be clear, I have not looked into the details of this myself.  I cannot guess how the actual performance would be impacted, but I'm also not entirely sure how important the ultimate performance for any single FFT is for real workflows.  That is, gridrec seems "damn fast" to me, and I would guess that preprocessing steps and managing multi-cores would have a non-zero impact on total runtime (so that small performance loss in the FFT may go undetected).  I do not have actual data to back that up, and I definitely cannot judge how important FFT speed is compared to ease of portability and maintenance.   But it might be worth checking out.

**Comment by carterbox** [comment]
@jrmadsen Thanks for the hints. I will put this on my list of things to do this week, but if I run into trouble, it will have to wait until August 18 because I'm attending a wedding and conferences.

**Comment by carterbox** [comment]
I published some commits, but I haven't tested them yet because I'm having conda troubles. I will copy them over to jrmadsen/tomopy when I have completed the tests.

**Comment by carterbox** [comment]
I tested on Linux only. It is now possible to build without MKL. In such a case, `gridrec` will exit and print a helpful error message. Tests will not complete with the no MKL build because the `gridrec` error is not a Python error. If we ever replace `ctypes` with `pybind11`, then we can use a C++ exception instead of `exit()`.

**Comment by carterbox** [comment]
Actually, there is some interaction between PTL and MKL that I don't understand. For some reason, I made it not possible to build with both PTL and MKL simultaneously.

**Comment by ssomnath** [comment]
Hello @jrmadsen and @carterbox. I would like to build tomopy for multi-GPU reconstruction without MKL for an IBM Power 9 Linux machine. For obvious reasons, I cannot use the X86 conda channels and have been building the dependencies manually by following [these instructions](https://gist.github.com/lbluque/1bed77425fc531a882cfcca68dd3cb6e).  I am very new to tomopy and would appreciate your guidance in building tomopy. Do I start from @jrmadsen 's fork as is or would I need to modify any of the build scripts?

**Comment by carterbox** [comment]
@ssomnath Please reach out to [Lukas Vlcek at ORNL](https://www.ornl.gov/staff-profile/lukas-vlcek). He is already working on porting TomoPy to POWER9, and we had an email exchange a few months ago about he should proceed with that work. Also, (as far as I know) @jrmadsen 's fork has no POWER9 related features, so basing your work on his unmerged commits is not advised.

That gist is out of date as of version 1.4(?) for various reasons. Our build system is now `scikit-build` and CMake based, and it builds like you would expect a python project to build. @jrmadsen did a good job of making all of the build options self-documented when passing the `--help` option to `setup.py`. i.e.

```bash
python setup.py build --help
```

Anyways, the issue of having too-many strict dependencies is one of the things a student intern will resolve this summer.

**Comment by ssomnath** [comment]
Thank you very much @carterbox. I am actually taking over Lukas' work and coincidentally received a copy of your communication with him today. 

We are not looking for a POWER9 optimized, just one that allows for multi GPU but doesn't require MKL as a hard dependency. We will pass on any feedback with regards to building (and optimizing) on POWER. Yes, I did see that you are hoping to recruit a summer student for refactoring and modularizing tomopy. 

For now, I started from the master branch and [put back](https://github.com/ssomnath/tomopy/commit/b55da98c2195890b59dde9ad44605d3b1d8322f7) the MKL / FFTW conditions from 1.4.x in `gridrec.c` that were removed in 1.5.0 in my fork.  It is my understanding that I need to look into the installation scripts to include FFTW.
